src/consul.rs

```diff
@@ -429,51 +436,101 @@ impl ConsulSession {
 /// # Background task ensuring own config key exists
 ///
 /// Makes sure that the config key we created still exists by long polling. If getting it fails for
-/// whatever reason we trigger the parent [`CancellationToken`] to shut down since we can no longer
-/// be sure that we have a valid locked session.
+/// whatever reason we trigger the parent [`CancellationToken`] to cancel since we can no longer be
+/// sure that we have a valid locked session. If the key exists but locked by the wrong session we
+/// also trigger a cancellation.
 async fn config_handler(
     client: ConsulClient,
-    mut peer_url: Url,
+    session_id: Uuid,
+    peer_url: Url,
     config_token: CancellationToken,
     parent_token: CancellationToken,
 ) {
-    // Perform stale reads when checking for our config.
+    // Consul documents that stale results are generally consistent within 50 ms, so let's sleep
+    // for that amount of time before we start checking to try to prevent spurious errors returned.
     //
-    // When the Consul server which is the Raft leader is restarted all KV reads by default return
-    // 500 errors until a new Raft leader is elected. For our usecase it's fine if the read value
-    // is a bit stale though.
-    peer_url.query_pairs_mut().append_pair("stale", "1");
+    // We still perform 5 retries at 1 second intervals since the maximum delay is theoretically
+    // unbounded.
+    tokio::time::sleep(Duration::from_millis(50)).await;
 
+    let mut failed_fetches = 0;
     let mut index = None;
+
+    let mut interval = interval(Duration::from_secs(1));
+    interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Skip);
     loop {
+        // Wait for an interval tick to have passed to prevent us from hammering the server with
+        // requests.
+        tokio::select! {
+            _ = config_token.cancelled() => {
+                trace!("Consul config handler was cancelled");
+                break;
+            },
```

Comment:
Do we really need to check for cancelled here if we're already checking immediately again for this down below?

Reply:
Good point, we _just don't have the time_.